keras/fit_scipy.py

```diff
@@ -0,0 +1,119 @@
+"""
+Fit a Keras model with any method from scipy.optimize.minimize.
+
+fit_scipy(..) is the only function which needs to be called externally.
+"""
+
+from __future__ import division
+import numpy as np
+import scipy as sp
+import scipy.optimize
+import keras
+from keras import backend as K
+
+
+def pack_theta(trainable_weights):
+    """ Flattens a set of shared variables (trainable_weights)"""
+    x = np.empty(0)
+    for t in trainable_weights:
+        x = np.concatenate((x, K.get_value(t).reshape(-1)))
+    return x
+
+
+def unpack_theta(model, theta):
+    """ Converts flattened theta back to tensor shapes of Keras model params """
+    weights = []
+    idx = 0
+    for layer in model.layers:
+        layer_weights = []
+        for param in layer.get_weights():
+            plen = np.prod(param.shape)
+            layer_weights.append(
+                np.asarray(
+                    theta[
+                        idx:(
+                            idx +
+                            plen)].reshape(
+                        param.shape),
+                    dtype=np.float32))
+            idx += plen
+        weights.append(layer_weights)
+    return weights
+
+
+def set_model_params(model, theta):
+    """ Sets the Keras model params from a flattened numpy array of theta """
+    trainable_params = unpack_theta(model, theta)
+    for trainable_param, layer in zip(trainable_params, model.layers):
+        layer.set_weights(trainable_param)
+
+
+def get_cost_grads(model):
+    """ Returns the cost and flattened gradients for the model """
+    trainable_params = get_trainable_params(model)
+
+    cost = model.model.total_loss
+    grads = K.gradients(cost, trainable_params)
+
+    return cost, grads
+
+
+def flatten_grads(grads):
+    """ Flattens a set tensor variables (gradients) """
+    x = np.empty(0)
+    for g in grads:
+        x = np.concatenate((x, g.reshape(-1)))
```

Comment:
Hi! Thanks for this PR. Saved me some time.
I'm using GPU optimized Theano as my backend, which means that `g` on this line is a `CudaNdarray` type object. Executing `g.reshape(-1)` raises the exception:

```
*** ValueError: size must remain unchanged, changed from 2000 to -1
```

Changing this to `g.reshape(g.size)` fixes the issue for me. I haven't investigated the problem further, but maybe others will run into this as well. You might want to change this before the PR is merged?